File: 168_ValidParentheses.cpp
/*
Problem: Check for Balanced Parentheses
Description: Check if string has balanced parentheses - (), {}, [].

Approach: Use stack. Push opening brackets. For closing, check if matching opening is on top.
*/

#include <bits/stdc++.h>
using namespace std;

bool isValid(string s) {
    stack<char> st;
    
    for (char c : s) {
        if (c == '(' || c == '{' || c == '[') {
            st.push(c);
        } else {
            if (st.empty()) return false;
            
            char top = st.top();
            if ((c == ')' && top == '(') ||
                (c == '}' && top == '{') ||
                (c == ']' && top == '[')) {
                st.pop();
            } else {
                return false;
            }
        }
    }
    
    return st.empty();
}

int main() {
    string s1 = "()[]{}";
    cout << s1 << " is " << (isValid(s1) ? "valid" : "invalid") << endl;
    // Output: valid
    
    string s2 = "([)]";
    cout << s2 << " is " << (isValid(s2) ? "valid" : "invalid") << endl;
    // Output: invalid
    
    return 0;
}

/*
Time Complexity: O(n)
Space Complexity: O(n)
*/

File: 168_ValidParentheses.py
"""
Problem: Check for Balanced Parentheses
Description: Check if string has balanced parentheses - (), {}, [].

Approach: Use stack. Push opening brackets. For closing, check if matching opening is on top.
"""

def isValid(s: str) -> bool:
    stack = []
    mapping = {')': '(', '}': '{', ']': '['}
    
    for c in s:
        if c in '({[':
            stack.append(c)
        else:
            if not stack or stack[-1] != mapping[c]:
                return False
            stack.pop()
    
    return len(stack) == 0

def main():
    s1 = "()[]{}"
    print(f"{s1} is {'valid' if isValid(s1) else 'invalid'}")
    # Output: valid
    
    s2 = "([)]"
    print(f"{s2} is {'valid' if isValid(s2) else 'invalid'}")
    # Output: invalid

if __name__ == "__main__":
    main()

"""
Time Complexity: O(n)
Space Complexity: O(n)
"""
